### PR TITLE
Feature karma https webserver

### PR DESCRIPTION
**Pull request by KrekkieD**
Specifies 2 new config options: `protocol` and `httpsServerOptions` to allow Karma web server to run on https if desired. Default is still http.

### CODE REVIEW

File: lib/executor.js [hunk 1]
```diff
@@ -10,7 +10,7 @@ var Executor = function (capturedBrowsers, config, emitter) {
     var nonReady = []
 
     if (!capturedBrowsers.length) {
-      log.warn('No captured browser, open http://%s:%s%s', config.hostname, config.port,
+      log.warn('No captured browser, open ' + (config.protocol || 'http:') + '//%s:%s%s', config.hostname, config.port,
```
Comment (Dignifiedquire) [inline lib/executor.js]: Please move the default setting into `lib/config.js`, so we don't have to add these checks everywhere

**Comment by KrekkieD** [comment]
@Dignifiedquire I have made adjustment for deeper integration. The config value is not available in the unit tests as they test some classes directly without going through config. Changed implementation to use function arguments where possible, and updated unit tests accordingly. Feels more in line with how the app is set up.

**Comment by KrekkieD** [comment]
Should [this line](https://github.com/KrekkieD/karma/blob/feature-karma-https-webserver/lib/reporter.js#L24) be updated to `'http[s]?:\\/\\/...'`? I am not sure what this function does, but it's the last hardcoded occurrence of `http:`.

File: lib/config.js [hunk 2]
```diff
@@ -118,6 +118,13 @@ var normalizeConfig = function (config, configFilePath) {
   // normalize urlRoot
   config.urlRoot = normalizeUrlRoot(config.urlRoot)
 
+  config.protocol = config.protocol || 'http:'
+  if (['http:', 'https:'].indexOf(config.protocol) === -1) {
```
Comment (Dignifiedquire) [inline lib/config.js]: Let's make the values `http` and `https` without the `:`, and strip it, if someone adds it by accident.

**Comment by Dignifiedquire** [comment]
> Should this line be updated to 'http[s]?:\\/\\/...'? I am not sure what this function does, but it's the last hardcoded occurrence of http:.

Yes this needs to be updated as well

File: lib/config.js (hunk 2 above)
Comment (KrekkieD) [inline lib/config.js]: How about we turn that around; add the `:` when omitted? Reason being that [NodeJS's url.parse](https://nodejs.org/api/url.html#url_url) returns the protocol including `:`, and it is used the same way by `lib/middleware/proxy.js`. If the property includes the colon it is more consistent and allows further use of the `url` class where desired.

Comment (Dignifiedquire) [inline lib/config.js]: yeah I'm okay with that

**Comment by KrekkieD** [comment]
Changes made. The travis build seems a bit flaky but I don't think it's caused by my changes.

File: lib/config.js [hunk 3]
```diff
@@ -118,6 +118,14 @@ var normalizeConfig = function (config, configFilePath) {
   // normalize urlRoot
   config.urlRoot = normalizeUrlRoot(config.urlRoot)
 
+  // force protocol to end with ':'
+  config.protocol = (config.protocol || 'http').split(':')[0] + ':'
+  if (config.protocol.match(/https?:/) === null) {
+    log.warn('"%s" is not a supported protocol, defaulting to "http:"',
+      config.protocol)
+    config.protocol = 'http:'
+  }
```
Comment (Dignifiedquire) [inline lib/config.js]: Could you please add a test for the above logic in `test/unit/config.spec.js`

File: lib/web-server.js [hunk 4]
```diff
@@ -61,7 +63,14 @@ var createWebServer = function (injector, emitter) {
...
 
-  var server = http.createServer(handler)
+  var serverClass = http
+  var serverArguments = [handler]
+
+  if (config.protocol === 'https:') {
+    serverClass = https
+    serverArguments.unshift(config.httpsServerOptions || {})
+  }
+  var server = serverClass.createServer.apply(null, serverArguments)
```
Comment (Dignifiedquire) [inline lib/web-server.js]: Could you please add a test for this, making sure when `https` is set, it's actually used

**Comment by Dignifiedquire** [comment]
Needs rebase onto latest master

**Comment by KrekkieD** [comment]
Tests added and pulled in latest master.

**Comment by Dignifiedquire** [comment]
Thanks @KrekkieD now squash it all together and we can get this shipped :)

**Comment by KrekkieD** [comment]
I believe, after some Googling on how to do that properly after pulling in commits from master, it is done ;)

**Comment by Dignifiedquire** [comment]
Thanks :octocat:

**Comment by Dignifiedquire** [comment]
Released in [v0.13.5](https://github.com/karma-runner/karma/releases/tag/v0.13.5)

**Pull request merged**